text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1124 @@
+  constitutes a breaking change between versions.
+
+- Telemetry schemas are identified by Schema URLs.
+
+- Telemetry sources (e.g. instrumentation libraries) should include a schema URL
+  in the emitted telemetry.
+
+- Telemetry consumers should pay attention to the schema of the received
+  telemetry. If necessary, telemetry consumers may transform the telemetry data
+  from the received schema version to the target schema version as expected at
+  the point of use (e.g. a dashboard may define which schema version it
+  expects).
+
+- OpenTelemetry will publish a telemetry schema as part of the specification.
+  The schema will contain the list of transformations that semantic conventions
+  undergo. The schema will be available to be referred and downloaded at well
+  known URLs.
+
+- OpenTelemetry instrumentation libraries will include the OpenTelemetry Schema
+  URL in all emitted telemetry.
+
+- OTLP will be modified to allow inclusion of a schema URL in the emitted
+  telemetry.
+
+- Third-party libraries, instrumentation or applications will be advised to
+  define and publish their own telemetry schema if it is completely different
+  from OpenTelemetry schema (or use OpenTelemetry schema) and include the schema
+  URL in the emitted telemetry.
+
+## What is Out of Scope
+
+- The concept of schema defined in this proposal does not attempt to fully
+  describe the shape of telemetry. The schema for example does not define all
+  possible valid values for attributes or expected data types for metrics, etc.
+  It is not a goal. Our goal is narrowly defined to solve the following problem
+  only: to allow OpenTelemetry Semantic Conventions to evolve over time. For
+  that reason this document is concerned with _changes_ to the schema as opposed
+  to the _full state_ of the schema. We do not preclude this though: the schema
+  file format is extensible and in the future may allow defining the full state
+  of the schema, see [Current State in Schema](#current-state-in-schema) in
+  Future Possibilities section).
+
+- We intentionally limit the types of transformations of schemas to the bare
+  minimum that is necessary to handle the most common changes that we believe
+  OpenTelemetry Semantic Conventions will require in the near future. More types
+  of transformations [may be proposed](#other-transformation-types) in the
+  future. This proposal does not attempt to support a comprehensive set of
+  possible transformation types that can handle all possible changes to schemas
+  that we can imagine. That would be too complicated and very likely
+  superfluous. Any new transformation types should be proposed and added in the
+  future to the schema file format when there is an evidence that they are
+  necessary for the evolution of OpenTelemetry.
+
+## Use Cases
+
+This section shows a couple interesting use-cases for the telemetry schemas
+(other uses-cases are also possible, this is not an exhaustive list).
```

Comment:
Regarding the use case where the backend is not aware of schemas and the collector is not used: should there be any SDK functionality for this use case, or is it left to each individual exporter to implement schema conversions?

Reply:
It is a possibility. We will need to decide if we want to put that burden on each language SDK vs implementing schema conversion once in the Collector.

Reply:
It would indeed be a burden on languages SDKS, however, I think when not providing this feature on SDK level we end up with inconsistent and unexpected user experiences.

E. g., the Jaeger and Zipkin exporters which every SDK ships by default wouldn't provide any schema mapping capabilities.

Apart from that I really like this proposal and the direction it's heading towards.

Reply:
I would vote against re-implementing schema conversion in every language.

Reply:
That would require everyone to use the collector?

Reply:
>That would require everyone to use the collector?

Either use a Collector or use a backend which understands schemas.

Alternatively, if there is no Collector and the backend does not understand schemas the system will operate just as it does today, without being aware of schemas. That is not a catastrophic situation, only things that rely on the schema may be broken, the rest should work. So the degraded behaviour in systems that do not understand schemas is no worse than we have today.